### PR TITLE
feat/multi_mapping_apis draft v0_1 for mulit map apis

### PR DESCRIPTION
**Pull request by wingleungchoi**
@ctlong `don't` read it. :smirk:  Let discuss with your plan tomorrow

### CODE REVIEW

File: lib/address_geocoder/MapApiManager [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Map API Manager don't Know HOW to do every action
+# know WHAT to do and WHO to do
+# know address_geocoder requirement
+
+require 'map_apis/*'
```
Comment (ctlong) [inline lib/address_geocoder/MapApiManager]: Should really only require the map_api that the user has chosen (can require in method)
e.g. `require "map_apis/#{api}"

File: lib/address_geocoder.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ class AddressGeocoder # :nodoc:
   CYCLEWITHPOSTAL   = { all: 1, remove_street: 2, remove_city: 3, remove_state: 4 }.freeze
   CYCLEWITHNOPOSTAL = { all: 5, remove_street: 6, remove_city: 7 }.freeze
 
-  attr_accessor :api_key, :country, :state, :city, :postal_code, :street, :language
+  attr_accessor :api_key, :country, :state, :city, :postal_code, :street, :language, :map_apis
```
Comment (ctlong) [inline lib/address_geocoder.rb]: Change name to `api`? `selected_api`? `chosen_api`? `api_service`?

File: lib/address_geocoder.rb [hunk 3]
```diff
@@ -30,46 +30,32 @@ def initialize(opt = {})
...
+    # 1. If address values have changed call third_party_api
+    call_map_apis if values_changed?
     # 2. If response failed return false
     return false unless @response.success?
     # 3. Initialize refined_address
     country_wo_postal = match_country.reject { |k| k == :postal_code }
     refined_address = { country: country_wo_postal, city: nil, state: nil, postal_code: nil, street: nil }
-    # 4. Pass refined address and google response to parser
+    # 4. Pass refined address and third_party_api response to parser
     parser = Parse.new(@response.result['results'][0]['address_components'], refined_address)
```
Comment (ctlong) [inline lib/address_geocoder.rb]: Unless we standardize results completely we're going to have to make a new parser method

File: lib/address_geocoder/MapApiManager [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  class MapApiManager
+    def initialize(opt = {})
+      # add attributes
+      # also validation inputs
+    end
+
+    def process
+      # call MapApiWrappers
+      # call ReportWriter
+      # response a result in a format of address_geocoder expected
```
Comment (ctlong) [inline lib/address_geocoder/MapApiManager]: Oh okay so it standardizes response

File: lib/address_geocoder/MapApiManager [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+class AddressGeocoder
+  class MapApiManager
+    def initialize(opt = {})
+      # add attributes
+      # also validation inputs
+    end
+
+    def process
+      # call MapApiWrappers
+      # call ReportWriter
```
Comment (ctlong) [inline lib/address_geocoder/MapApiManager]: Right before here we are going to have to parse because I doubt that we can compare results without parsing

Comment (ctlong) [inline lib/address_geocoder/MapApiManager]: Also would only call this if multiple apis specified correct?

**Comment by wingleungchoi** [comment]
@ctlong  i will discuss in the morning. I briefly read another gem you recommend, the structure is much better and more modularised.
How do we compare the result of different APIs result, how to structure this logics in our gem will be fun and a good lesson for us. :sunglasses:

File: lib/address_geocoder.rb (hunk 2 above)
Comment (wingleungchoi) [inline lib/address_geocoder.rb]: yes

File: lib/address_geocoder/MapApiManager (hunk 1 above)
Comment (wingleungchoi) [inline lib/address_geocoder/MapApiManager]: this is cool. it would be faster when require api only if it is faster.

**Pull request closed without merging**